### ISSUE TITLE
[Needs regression test steps] [$1000] When copy pasting message containing email/links, email address/link occupies new line.

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#9250]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Typed a message [a@mail.com](mailto:a@mail.com) is my email. [b@mail.com](mailto:b@mail.com) is someone else's email.
Copied and pasted in the text box below

## Expected Result:
Text should copy in the same line 

## Actual Result:
When copy pasting message containing email/links, email address/link occupies new line.

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- Desktop App

**Version Number:** 1.1.69-2
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
<img width="616" alt="Screenshot 2022-05-26 at 12 19 53 PM" src="https://user-images.githubusercontent.com/43995119/171100817-db8c5fb8-65dc-4052-ab1c-e18b581ab557.png">

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @MonilBhavsar 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1653548218883739

**Job Post:** https://www.upwork.com/jobs/~01e09832b6e53bb854

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Julesssss** [comment]
Reproduced, this should be fixed.

**Comment by luacmartins** [comment]
Hmm I wonder why I got assigned if @Julesssss is an ECM and is still assigned as well 🤔 @Julesssss should I keep myself assigned to this one?

**Comment by Julesssss** [comment]
Odd, I was OOO but I didn't think auto-assignment was smart enough to know that? Anyway, I'll keep it 👍

**Comment by Tushar-Beladiya** [comment]
If I want to contribute and I want to fixed this issue, then Can I do this?

**Comment by parasharrajat** [comment]
You can do this. Please share your proposal.

**Comment by NehaSantani** [comment]
Proposals :

Hey,
 There is a method in CopySelectionHelper.js file
  ```
   copySelectionToClipboard() {
        const selectionMarkdown = SelectionScraper.getAsMarkdown();
        Clipboard.setString(selectionMarkdown);
    }

```
Above method is converting the copied text to HTML then to  markdown.

There are two options we can have to solve this issue:

1.) We can update the **getHtmlOfSelection()** for selected text and convert it into markdown as below

 ```
const getHTMLOfSelection=()=>{
    if (window.getSelection) {
        const selection = window.getSelection();
        const parser = new ExpensiMark();
        return parser.replace(selection);  
     }

      // If browser doesn't support Selection API, returns empty string.
      return '';
   
}
 ```    
since, parser.replace() method is already being used in while we are trying to edit a comment, we can use the same method for selection !

2.) We can set the **selected text in clipboard as a plain text** which is being selected by updating **copySelectionToClipboard()** method as below.
 ```
 copySelectionToClipboard() {
       if(window.getSelection())
        Clipboard.setString(window.getSelection().toString());
    }
 ```
or by updating current **getHTMLOfSelection()** as returning
```
             return div.innerText;
```
instead of 
```
            return div.innerHTML;
```

**Comment by parasharrajat** [comment]
Hey @NehaSantani, thanks for the proposal. Did you test this solution on local setup? 

How will you convert the text to Markdown.

**Comment by NehaSantani** [comment]
Hi @parasharrajat,
The text is being converted into MarkDown in the similar manner like it is currently there using below method 


>**index.js in /SelectionScraper.js**
```
const getAsMarkdown = () => {
    const selectionHtml = getHTMLOfSelection();

    const domRepresentation = parseDocument(selectionHtml);
    domRepresentation.children = _.map(domRepresentation.children, c => replaceNodes(c));

    const newHtml = render(domRepresentation);

    const parser = new ExpensiMark();

    return parser.htmlToMarkdown(newHtml);
};

```
which inturn is called from below method

```
 copySelectionToClipboard() {
        const selectionMarkdown = SelectionScraper.getAsMarkdown();
        Clipboard.setString(selectionMarkdown);
    }
```
_The whole conversion currently in the code is going like this
(selectedText -----> HTML ------> Markdown)_

However, the issue is inside HTML conversion of text!
Actual Issue: The code done inside getHTMLOfSelection() method is returning HTML with some default style tags which is appending it to next line after converting into markdown.

The only change we need to do to solve the issue mentioned is inside **getHTMLOfSelection()** method which is called from getAsMarkdown() method.

**Updated getHTMLofSelection()**.

> **index.js in /SelectionScaper.js**


```
const getHTMLOfSelection = () => {
    if (window.getSelection()) {
    const selection = window.getSelection().toString();
    const parser = new ExpensiMark();
    return parser.replace(selection);  
 }  // If browser doesn't support Selection API, returns empty string.
 return '';
}
```

My proposal to use above code in getHTMLOfSelection() method is because we are already using parser.replace to convert it to html while editing a message!

Also,Yes i tested it in local.
attached is the Recording for the same.


https://user-images.githubusercontent.com/4989101/172761596-2c5bcda3-d33b-4678-ad6e-709e18ed2a2f.mov

**Comment by parasharrajat** [comment]
Try this to copy chat between two messages. Start from bottom to the third last line(Hi Neha) of the previous message. Then paste it on composer and let me know what output you get.

**Comment by NehaSantani** [comment]
This way?

https://user-images.githubusercontent.com/4989101/172856353-e9ac2ece-fa0c-4536-9c9f-bd112141eebb.mov

**Comment by NehaSantani** [comment]
https://user-images.githubusercontent.com/4989101/172863073-be2f777e-7ea0-46bc-8eef-9be4fa349e80.mov


Also, if you have any specific text to be tested with styles(bold,italic,quotes) you can send me!
I will record for that aswell!

**Comment by parasharrajat** [comment]
Ok, thanks for sharing. I am amazed to see that it is working. How is your function `getHTMLOfSelection` converts elements with `data-testid` to proper tags? 

> const getHTMLOfSelection = () => {
    if (window.getSelection()) {
    const selection = window.getSelection().toString();
    const parser = new ExpensiMark();
    return parser.replace(selection);  
 }  // If browser doesn't support Selection API, returns empty string.
 return '';
}

window.getSelection().toString() gives string neither markdown nor HTML. Then how are you getting HTML. `parser.replace(selection);` converts MD to HTML.

**Comment by Julesssss** [comment]
Not overdue

**Comment by marktoman** [comment]
### Proposal

HTML generated by the app (meaning HTML rendered in the browser) for comment links is not the anchor element alone, but the anchor nested inside a div tree.

When we copy, we copy from the rendered HTML, and while the copy strips and skips unneccessary parts, it preserves the anchor tree. Because of that, even inline anchors (rendered inline because of styling) are then converted with line breaks because the div element is a block element.

Comment input text:
```
test.com test

test.com
test
```

When submitted, selected, and copied, it produces the following:

Copied HTML ([here](https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L139)):
```html
<comment><div><div><div><a href="https://test.com">test.com</a></div></div></div><span> test</span><br>
<br>
<div><div><div><a href="https://test.com">test.com</a></div></div></div><br>
<span>test</span><br></comment>
```

MD generated from it:
```md
[test.com](https://test.com)
 test

[test.com](https://test.com)

test
```

There is nothing wrong with the conversion itself. We have to avoid producing the div tree on copy.

We mark the component tree that makes the comment link by wrapping it in an inline element with an attribute.
Change this line:
https://github.com/Expensify/App/blob/5ce1b2cedc08dddd0f8dea4a38be0fcce933d05e/src/components/AnchorForCommentsOnly/index.js#L33
```diff
- return <BaseAnchorForCommentsOnly {...propsToPass} />;
+ return Platform.select({
+     web: <span datatype='link'><BaseAnchorForCommentsOnly {...propsToPass} /></span>,
+     default: <BaseAnchorForCommentsOnly {...propsToPass} />
+ });
```

Then, we unwrap them when the user copies a selection:
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L81
```diff
+ const getNestedAnchor = (element) => {
+     if (element.name === 'a') {
+         return element;
+     }
+     if (element.children && element.children.length === 1) {
+         return getNestedAnchor(element.children[0]);
+     }
+ }
```
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L97
```diff
+ if (dom.attribs && dom.attribs.datatype === 'link') {
+     const a = getNestedAnchor(dom);
+     if (a) {
+         domName = 'a';
+         domAttribs.href = a.attribs.href;
+         domChildren = [{
+             type: 'text',
+             data: Str.htmlEncode(a.children[0].data)
+         }];
+     }
+ }
```
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L116
```diff
- if (dom.children) {
+ if (!domChildren && dom.children) {
```


HTML cloned when copying ([here](https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L139)):
```html
<comment><a href="https://test.com">test.com</a><span> test</span><br>
<br>
<a href="https://test.com">test.com</a><br>
<span>test</span><br></comment>
```

MD generated from it:
```md
[test.com](https://test.com) test

[test.com](https://test.com)
test
```

Alternatively, instead of marking the anchor tree explicitly, we can just take `<comment>...</comment>` as the root and unwrap all anchors directly to it. Not sure which is better.

**Comment by marktoman** [comment]
# Proposal

Implementation of the idea mentioned previously. This way, we don't need to wrap the anchor tree in `<span>`. We identify it by the fact that it is rooted in `<comment>` and each element has a single child and the tree ends with the anchor. The wrapping elements are meaningless in every way for the copy (see Copied HTML above).

https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L81
```diff
+ const getNestedAnchor = (element) => {
+     if (element.name === 'a') {
+         return element;
+     }
+     if (element.children && element.children.length === 1) {
+         return getNestedAnchor(element.children[0]);
+     }
+ }
```
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L92
```diff
+ const parent = dom.parentNode;
+ const grandParent = parent && parent.parentNode;
+ 
+ if (parent && parent.attribs && parent.attribs[tagAttribute] === 'comment' ||
+     grandParent && grandParent.attribs && grandParent.attribs[tagAttribute] === 'comment') {
+     const a = getNestedAnchor(dom);
+     if (a) {
+         return {
+             ...dom,
+             name: 'a',
+             attribs: a.attribs,
+             children: [{
+                 type: 'text',
+                 data: Str.htmlEncode(a.children[0].data)
+             }],
+         };
+     }
+ }
```

**Comment by parasharrajat** [comment]
Let's say the user does not copy the whole message including the comment node. Will this work? Whether the dom will contain a comment node as well. 

Why are we checking parent or grandParent to be comment? What is the reason(logic) behind that?

Is there a better way of detecting the extra divs? There are many utils available to manage dom with htmlparser lib. have you tried any? 

It would be great to reduce these divs for all types of tags.

> Remove the nested divs whose only child is either another div or a content-node. And keep only the content-node. 
Content-node: Any node that has semantic meaning in HTML and contains the content node (Text).

E.g.

```html
<comment><div><div><div><a href="https://test.com/">test.com</a></div></div></div><span> test</span><br>
<br>
```
will become
```
<comment><a href="https://test.com/">test.com</a><span> test</span><br>
<br>
```

**Comment by marktoman** [comment]
# Proposal 
Yes, we can reduce it the similar way `htmlToMarkdown` does. I was trying to preserve the original structure, but there may be no good need for it.

First, let's add a modified version of `replaceBlockWithNewLine` from `ExpensiMark`. The only difference is that we use `<br>` for the new line.

Add it here:
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L128
```js
 const replaceBlockWithNewLine = (htmlString) => {
    const splitText = htmlString.split('[block]');
    let joinedText = '';

    // Delete whitespace at the end
    while (splitText.length) {
        if (splitText[splitText.length - 1].trim().length > 0 || splitText[splitText.length - 1].match(/\n/)) {
            break;
        }
        splitText.pop();
    }

    splitText.forEach((text, index) => {
        if (text.trim().length === 0 && !text.match(/\n/)) {
            return;
        }

        // Insert '\n' unless it ends with '\n' or '>' or it's the last [block].
        if (text.match(/[\n|>][>]?[\s]?$/) || index === splitText.length - 1) {
            joinedText += text;
        } else {
            joinedText += `${text}<br>`;
        }
    });

    return joinedText;
};
```

Next, we take the definition of block elements and leave out headings from this line:
https://github.com/Expensify/expensify-common/blob/3d87e9b87c6e5ce5fa4bac05bceb3cdf3ad6ca56/lib/ExpensiMark.js#L459

Change
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L143
```diff
-     return parser.htmlToMarkdown(newHtml);
+    const rawHtml = newHtml.replace(/<div.*?>|<\/div>|<comment.*?>|\n<\/comment>|<\/comment>/gm, '[block]');
+    const cleanHtml = replaceBlockWithNewLine(rawHtml);
+    return parser.htmlToMarkdown(cleanHtml);
```

**Comment by marktoman** [comment]
This works with both partial selection and selection spanning across multiple comments.

**Comment by parasharrajat** [comment]
I am not able to understand the logical base on the code above. It can easily break with changes to HTML structure. It seems that it is already present on the common repo so ignore it.

It seems like a regression from https://github.com/Expensify/App/pull/8743. cc: @rushatgabhane.

---

>  The only difference is that we use <br> for the new line.

What is the effect of this change? How does this work?

**Comment by rushatgabhane** [comment]
@ahmdshrif might be more familiar here. Let us know if you have a better solution

**Comment by marktoman** [comment]
The production change would be adding a new parameter to `replaceBlockWithNewLine` instead of copying it from the common repo. The parameter would allow a custom new line string. We need `\n` for MD and `<br>` for HTML.

> It can easily break with changes to HTML structure.

It removes all the block elements with no loss of information. This is what is currently used for MD, we just take advantage of it. You can compare HTML that goes in and out. The clean HTML is what we want to put in the clipboard as well.

> What is the effect of this change? How does this work?

It places `<br>` instead `\n` for the new line. When MD conversion uses the same function, it [strips HTML](https://github.com/Expensify/expensify-common/blob/3d87e9b87c6e5ce5fa4bac05bceb3cdf3ad6ca56/lib/ExpensiMark.js#L460) before calling it, so it loses `<br>` elements, which affects [when the new line character is placed](https://github.com/Expensify/expensify-common/blob/3d87e9b87c6e5ce5fa4bac05bceb3cdf3ad6ca56/lib/ExpensiMark.js#L423). We can focus on MD alone and avoid or change the HTML stripping, but producing clean HTML plays well with the other clipboard issue, which should use it.

**Comment by ahmdshrif** [comment]
>@ahmdshrif might be more familiar here. Let us know if you have a proposal

thanks @rushatgabhane  i will debug this . 

> The production change would be adding a new parameter to replaceBlockWithNewLine instead of copying it from the common repo. The parameter would allow a custom new line string. We need \n for MD and <br> for HTML.

the`replaceBlockWithNewLine`  only use on  `htmlToMarkdown`  so we expect to return MD only. 
and also takes text with no HTML and with specific shape so it's not general to use anyway. 

>It places <br> instead \n for the new line. When MD conversion uses the same function, it [strips HTML](https://github.com/Expensify/expensify-common/blob/3d87e9b87c6e5ce5fa4bac05bceb3cdf3ad6ca56/lib/ExpensiMark.js#L460) before calling it, so it loses <br> elements, which affects [when the new line character is placed](https://github.com/Expensify/expensify-common/blob/3d87e9b87c6e5ce5fa4bac05bceb3cdf3ad6ca56/lib/ExpensiMark.js#L423). We can focus on MD alone and avoid or change the HTML stripping, but producing clean HTML plays well with the other clipboard issue, which should use it.

any `<br>` will convert to new line `\n` before we use `replaceBlockWithNewLine`

**Comment by ahmdshrif** [comment]
so I do quick debug 
I log the HTMLString that we receive in `htmlToMarkdown` function 
and then render it on https://htmledit.squarefree.com/  (use any thing render html) 

so the result there is new line after any email  when render the html as shown 
so there is no issue on `htmlToMarkdown` or `replaceBlockWithNewLine`  
the issue with HTML itself 

![Screen Shot 2022-07-05 at 1 18 58 AM](https://user-images.githubusercontent.com/21364901/177224688-7e6e08fe-a7b8-4482-9e9d-d1fb4c4817b6.png)
 
### Proposal 

the issue now on html . 
so what i note we have the following 
```tsx
<div>
    <a href="test@test.com"> text </a>
</div>
```

and the new line comes from this extra div which we have many extra div that does not mean anything and sometimes break the code as in this case. 
so I will suggest filtering this extra div . and map only div that contain text inside it or more than one child. 

in https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L92 will add the following 

```ts 
    if (dom.name === 'div' && dom.children.length === 1 && dom.children[0].type !== 'text') {
        return replaceNodes(dom.children[0]);
    }
```
and will fix it and will filter all this extra div we do not needs. 

cc: @parasharrajat  @rushatgabhane

**Comment by parasharrajat** [comment]
Thanks, @ahmdshrif. I will review your proposal later today.

**Comment by kevinksullivan** [comment]
@parasharrajat did you get a chance to review the proposal?

**Comment by parasharrajat** [comment]
Oh, not yet. I will do that soon.

**Comment by marktoman** [comment]
# Proposal

I have tested this on the full MD set supported by the app, and the issue reveals the need for a more substantial fix.

The source is that dealing with block elements is split between the HTML and MD generators. The MD generator does not use full information about the DOM structure. The generated HTML should be unambiguous for the MD generator and other use cases.

See the full detail in the comments.

Add:
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L9
```diff
+ const specialBlockElements = ['comment'];
```

Remove `replaceNodes`:
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L82-L127
Add:
```js
// Determines if any parent node has tagAttribute that indicates content.
const hasParentWithContentTag = (node) => {
    if (!node) {
        return false;
    }

    // If the node has tagAttribute that should not be skipped and that is not a special block, then it is content.
    if (node.attribs
        && node.attribs[tagAttribute]
        && !elementsWillBeSkipped.includes(node.attribs[tagAttribute])
        && !specialBlockElements.includes(node.name)) {
        return true;
    }
    return hasParentWithContentTag(node.parent);
};

const replaceNodes = (node) => {
    if (node.type === 'text') {
        // Encoding HTML chars '< >' in the text, because any HTML will be removed in stripHTML method.
        const encodedText = Str.htmlEncode(node.data);

        // When we copy accross multiple lines from the rendered DOM, we get redundant \n text elements.
        if (encodedText === '\n') {
            return [];
        }

        // If the comment text contained formatting, the comment is converted to HTML on the server.
        // The comment-specific HTML is stored inside the data-testid attributes. All \n breaks are replaced with
        // <br> by the server.
        // If the comment is plain text, the entire text is preserved including the \n line breaks.
        // We need to replace them ourselves. Convert all lines, except the last line, unless it ends with \n.
        const lines = encodedText.split('\n');
        return _.chain(lines)
            .map((line, i) => (i < lines.length - 1 || (i === lines.length - 1 && encodedText.endsWith('\n'))
                ? [{...node, data: line}, new Element('br')]
                : {...node, data: line}))
            .flatten()
            .value();
    }

    // Resolve tagAttribute (i.e. data-testid) to the tag name.
    // We are skipping elements which have html and body in data-testid, since ExpensiMark can't parse it. Also this data
    // has no meaning for us.
    const resolvedNodeName = node.attribs && node.attribs[tagAttribute] && !elementsWillBeSkipped.includes(node.attribs[tagAttribute])
        ? node.attribs[tagAttribute]
        : node.name;

    // Skip images, which are not supported inside text comments.
    if (['svg', 'img'].includes(resolvedNodeName)) {
        return [];
    }

    // Replace and flatten the top-level children. Regular children are rooted inside each element and are not flattened.
    const replacedNodeChildren = _.chain(node.children).map(replaceNodes).flatten().value();

    // Replace inline containers with their children.
    if (resolvedNodeName === 'span') {
        return replacedNodeChildren;
    }

    // Always break lines for special blocks
    if (specialBlockElements.includes(resolvedNodeName)) {
        return [...replacedNodeChildren, new Element('br')];
    }

    // Decide if div brakes the line.
    if (resolvedNodeName === 'div') {
        // A parent with a certain tagAttribute indicates that this is a content tree, which means that
        // all block element inside it are redundant.
        if (hasParentWithContentTag(node)) {
            return replacedNodeChildren;
        }

        // Otherwise, any div directly containg text should break the line.
        if (_.some(node.children, c => c.type === 'text')) {
            return [...replacedNodeChildren, new Element('br')];
        }

        // If the div does not contain text, it is redundant - skip it.
        return replacedNodeChildren;
    }

    // We need to preserve href attribute in order to copy links.
    const nodeAttribs = node.attribs && node.attribs.href
        ? {href: node.attribs.href}
        : {};

    // Return all other nodes with stripped attributes and replaced children, otherwise as they are.
    return [{
        ...node,
        name: resolvedNodeName,
        attribs: nodeAttribs,
        children: replacedNodeChildren,
    }];
};
```
Replace the call:
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L137
```diff
- domRepresentation.children = _.map(domRepresentation.children, c => replaceNodes(c));
+ // Replace and flatten the top-level children. Regular children are rooted inside each element and are not flattened.
+ domRepresentation.children = _.chain(domRepresentation.children).map(replaceNodes).flatten().value();
```

### Examples

Add the examples below to test it.

![image](https://user-images.githubusercontent.com/47689923/178538288-0ef508eb-32b5-49e6-a51b-66fba3c7cc28.png)


Send this to a contact (plain text with lines)
```
line a
line b
line c
```

Send this from that contanct to yourself (links)
```
link.com inline

link.com
next line
```

Send this back to the contact (the rest)
````
> quote

```
code
```

*bold*
_italic_
~strikethrough~

text `inline code` text
````

Test it by selecting each comment, a subset of each comment, and across multiple comments. Observe both `newHtml` and the generated MD.

**Comment by parasharrajat** [comment]
ETA for review: 24 hours.

**Comment by parasharrajat** [comment]
Looking into the proposals now.

**Comment by kevinksullivan** [comment]
Hey @parasharrajat please provide an update once you've gotten around to testing. Thanks!

**Comment by Julesssss** [comment]
Awaiting update.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by ahmdshrif** [comment]
@mvtglobally I am still able to repro it on prod. 

https://user-images.githubusercontent.com/21364901/182325807-8b570f51-b10d-4a4b-96ab-c6325aa417fe.mov

@parasharrajat  what do you think about my proposal?

**Comment by Julesssss** [comment]
No change. Awaiting confirmation in further retesting

**Comment by Julesssss** [comment]
Hey @kevinksullivan, could we please double the amount?

**Comment by Julesssss** [comment]
Holding for proposal review.

**Comment by Julesssss** [comment]
> No need to double just yet. I was busy with some stuff, let me review the existing proposals. Sorry for the delay.

Hey @parasharrajat, do you have a moment to review these?

**Comment by Julesssss** [comment]
No update

**Comment by kevinksullivan** [comment]
Price doubled

**Comment by kevinksullivan** [comment]
Before doubling... @parasharrajat did you get a chance to review the proposals?

**Comment by parasharrajat** [comment]
Oh, this is getting delayed. I will review it soon. Updated proposals need to be tested but it seems that they are working as per expected behavior. So I would say don't double it yet and I will ping you when needed. Thanks for the bump.

**Comment by Julesssss** [comment]
Proposal still needs review.

**Comment by Julesssss** [comment]
The PR does seem to resolve the issue, so lets keep this on hold for now.

**Comment by arielgreen** [comment]
@kevinksullivan is this on hold? Can you please update the title if so?

**Comment by Julesssss** [comment]
I can't recall why this was held and can't figure it out by rereading all the comments in this thread. 

@Expensify/applauseleads would you mind retesting this.

**Comment by kevinksullivan** [comment]
It wasn't on hold, we were [waiting on review](https://github.com/Expensify/App/issues/9250#issuecomment-1270491967) originally, so it didn't seem like we should continue doubling if the existing proposal would solve the problem CC @parasharrajat

**Comment by parasharrajat** [comment]
Really confusing but I agree there is a proposal that needs review. I will pick this up tomorrow.

**Comment by kevinksullivan** [comment]
@parasharrajat what's the latest on this one? Did you get around to reviewing? Or should we double?

**Comment by parasharrajat** [comment]
Not yet. Switching among priorities ATM. I will take a look as soon as possible. Last time I reviewed this I was confused with some implementations done in the past. 

Steps need to be taken: 
1. Review the old PRs to determine why and what was implemented. This was a regression best of my knowledge. 
2. Review the two latest proposals. Test them for possible regressions.

**Comment by JmillsExpensify** [comment]
@Julesssss To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by Julesssss** [comment]
Hey @parasharrajat, we can swap out with another C+ if you have too much on at the moment. Let me know what suits you best 👍

**Comment by parasharrajat** [comment]
Sure. Let's do that.

**Comment by Julesssss** [comment]
Asked for C+ volunteers [here](https://expensify.slack.com/archives/C02NK2DQWUX/p1668771997922969)

**Comment by dylanexpensify** [comment]
@Julesssss given this issue is one of our oldest, and based on [this convo](https://expensify.slack.com/archives/C01SKUP7QR0/p1669075981370149?thread_ts=1669034215.991529&cid=C01SKUP7QR0), let's move this internally.

**Comment by ahmdshrif** [comment]
I posted proposals from more than  4 months and have not get reviewed until now!!! 
 https://github.com/Expensify/App/issues/9250#issuecomment-1174469689

**Comment by dylanexpensify** [comment]
@ahmdshrif sorry about that, it does indeed look like we've been waiting on your proposal to be reviewed! @aimane-chnaif can we get @ahmdshrif proposal reviewed and if acceptable as-is, we will move forward with hiring, etc cc @Julesssss

**Comment by dylanexpensify** [comment]
thanks @aimane-chnaif! Very appreciated!

**Comment by kevinksullivan** [comment]
@aimane-chnaif thanks for stepping in here. Do you have an ETA for when you will be done reviewing?

**Comment by marktoman** [comment]
@aimane-chnaif Please see the Examples part of my proposal, which are the cases that need to be handled. My solution was the only one satisfying them.

**Comment by ahmdshrif** [comment]
@aimane-chnaif  the current implementation is recursive so I go with the same to not break anything 
and I think recursive is the best way to implement DFS which is similar to our problem. 

https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L117

**Comment by ahmdshrif** [comment]
it's recursive.
recursive is a function call itself :D 
and here we call `replaceNodes(c)`
anyway let's check the regression but can you explain more about what is the problem with `>` or (```)

**Comment by ahmdshrif** [comment]
@aimane-chnaif ok it's look this block of code do this tages so we should's skip it by early return 
https://github.com/Expensify/App/blob/965c16c2bf4bdb878d8148302e435b0aabe9a13b/src/libs/SelectionScraper/index.js#L100-L109

instead, we can add this code  to else 
 ```diff
    if (dom.attribs && dom.attribs[tagAttribute]) {
...
+    } else if (dom.name === 'div' && dom.children.length === 1 && dom.children[0].type !== 'text') {
+        return replaceNodes(dom.children[0]);
+   }

```

this fixes the regression

**Comment by marktoman** [comment]
> @marktoman ok, can you explain all the cases handled? I will look into it further. And your solution breaks line at the end after copying test message above. Please update your proposal with that fix

@aimane-chnaif Which test message, [this one](https://github.com/Expensify/App/issues/9250#:~:text=Typed%20a%20message%20a%40mail.com%20is%20my%20email.%20b%40mail.com%20is%20someone%20else%27s%20email.%0ACopied%20and%20pasted%20in%20the%20text%20box%20below)? It works on my end.

Sure, let me elaborate on the cases needed to be handled.

**Comment by ahmdshrif** [comment]
@aimane-chnaif  yes because the blockquote is mapped to div with attribute `data-testid: 'blockquote'`
so we shouldn't skip the div that contains this attribute. 

it should be one of our checks but I add it to else to avoid any special attribute we look at it now or in the future.

**Comment by marktoman** [comment]
@aimane-chnaif Let me respond in a day. I remember more hidden problems around those edge cases (hence my complex solution), but I didn't have the time to go through it again yet.

**Comment by marktoman** [comment]
@aimane-chnaif I cannot replicate the regression. This is what I enter and get:
![image](https://user-images.githubusercontent.com/47689923/206094655-3ed89d83-6ec1-491b-86d0-23485ffc070a.png)
How does it look on your end?

**Comment by marktoman** [comment]
@aimane-chnaif 
No, you were right. I was copying a message rather than a selection, which works differently. See the updated solution.

So the problem is straightforward. I was working on another issue at the time, which was to provide rich types for the clipboard. Before, we only put MD-formatted text as a single mime type in the clipboard. I treated this issue as complimentary to provide reliable HTML for other apps and parsers, including our HTML to MD converter. Whenever content is nested in multiple divs, it is not deterministic how many line breaks they should be replaced by, so each app does it differently. Instead of handling each scenario case-by-case, which is unreliable long term, I came up with a generic solution. For line breaks, it means avoiding divs altogether and using only br where we determine a line break should be from the context. It also deals with all the remaining noise. Examples where this breaks things is copying over multiple messages (line breaks go out of hand in multiple apps), or the noise like empty image elements wrapped in divs. All the cases are described in detail in the solution. You can see the produced clipboard content by pasting it [here](https://evercoder.github.io/clipboard-inspector/).

**Comment by ahmdshrif** [comment]
@aimane-chnaif  Should we wait to review the new prosal from mark ?!! 
my proposal already works and we go with the first complete proposal as contributing guide says.

I also still say going to the top of the tree with every node is not correct from performance wise .

**Comment by Julesssss** [comment]
> we go with the first complete proposal as contributing guide says.

I don't think that is true, we go with the most ideal proposal based on a case-by-case basis. Having said that I'm still leaning towards this proposal due to simplicity at the moment.

**Comment by Julesssss** [comment]
So @marktoman, is [this](https://github.com/Expensify/App/issues/9250#issuecomment-1182002856) still your latest proposal? I'm struggling to see why this solution is preferable, to be honest. Partly due to complexity.

**Comment by marktoman** [comment]
@Julesssss Yes, please see https://github.com/Expensify/App/issues/9250#issuecomment-1342127627 where I elaborate.

**Comment by Julesssss** [comment]
Okay, thanks all. Let's go with the far simpler solution.

**Comment by kevinksullivan** [comment]
Here's a new job post since the previous one expired:

https://www.upwork.com/jobs/~019fadcb4f471fe494

@aimane-chnaif sent you an offer for C+, @ahmdshrif please apply. Thanks!

**Comment by shoaibarham** [comment]
Proposal:
Thank you for the opportunity to work on this project. I am an experienced React Native developer with a strong understanding of financial transactions and security.

Based on the GitHub issue, I understand that the goal of this project is to unify the front-end across platforms using React Native. To accomplish this, I propose the following steps:

First, I will review the current codebase to understand the existing architecture and identify any potential challenges or issues that may arise during the migration.

Next, I will create a fork of the codebase and create a new branch for my changes.

I will then begin implementing the necessary changes to migrate the front-end to React Native, ensuring that the code is well-structured, efficient, and maintainable.

As I make progress, I will periodically push my changes to the branch and create pull requests for review by the Expensify team. I will also provide screenshots and confirmation that I have tested the changes on all relevant platforms.

Once the changes have been reviewed and approved, I will merge the code into the main codebase and verify that the front-end is functioning correctly across all platforms.

I am confident that my technical expertise and experience with React Native will enable me to successfully complete this project and deliver high-quality code that meets your requirements. Thank you for considering my proposal, and I look forward to the opportunity to work with you.

**Comment by Zvukamar** [comment]
Proposal: 

In src/components/Composer/index.js

Line 222:

Change this:
const pastedHTML = event.clipboardData.getData(TEXT_HTML);

To this:
const pastedHTML = event.clipboardData.getData(TEXT_HTML).replace(/div/g, 'span');

Explain:
event.clipboardData.getData(TEXT_HTML) returns `<comment><div>abc</div><span>abc</span><div>abc</div</comment>`

In render each `<div>` by default get style of display: "block" and catch all the row,
because of that we have \n in plain text
If we replace all `<div>` elements with `<span>` elements all works good.

Tested all the cases above, everything seems good.

**Comment by ahmdshrif** [comment]
> @ahmdshrif you're assigned now. please feel free to raise PR.

I am working on it

> @ahmdshrif please apply. Thanks!

@kevinksullivan   done. Thanks all.

**Comment by trjExpensify** [comment]
Commented[ here](https://expensify.slack.com/archives/C01GTK53T8Q/p1670867982347419), but you should be able to get unblocked here now after whitelisting. 👍

**Comment by ahmdshrif** [comment]
I fill out the form.  and wait to add me to the whitelist.

**Comment by mallenexpensify** [comment]
Gotcha covered @aimane-chnaif , @ahmdshrif too!  
Thanks @aimane-chnaif for stepping in to get this over the finish line and for @ahmdshrif for being extra patient with us.  Thanks to the other contributors who submitted proposals and commented too!

@aimane-chnaif I did a quick check of the timeline but didn't factor in _down_ time with IP issues.  I'm seeing
- Hired Dec 9th
- Merged 16th

I'm seeing [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1670957921425999) on 12/13 about the IP issue, with the 10th and 11th being weekend days, it makes sense that y'all might be affected by the issue.  @aimane-chnaif and/or @ahmdshrif, if you were affected, can you provide some details/links and reasoning in a post so we can have it for _audit_ purposes? It doesn't have to be detailed and we only need one of you to do it. Thx

**Comment by mallenexpensify** [comment]
Thanks for the breakdown @aimane-chnaif, I agree.  I have issued 50% bonuses to you and @ahmdshrif . 

Leaving open for regression test steps/BZ check list

**Comment by kevinksullivan** [comment]
Still climbing through the OOO backlog, will knock this out tomorrow AM.
